clippy.toml

```diff
@@ -1,3 +1,4 @@
 disallowed-methods = [
   { path = "futures::channel::mpsc::unbounded", reason = "does not enforce backpressure" },
 ]
+type-complexity-threshold = 430
```

Comment:
For which type is this needed?

Reply:
By the way, the default is 250. I have no idea what the units are here. 430 was found experimentally.
This only matters when doing non-custom, out-of-the box `cargo clippy` that I like to do locally.
There's a handful of generic type parameters that are themselves types with generic type parameters, tuples involved, that trigger this.

````
warning: very complex type used. Consider factoring parts into `type` definitions
  --> swarm/src/connection/pool/task.rs:62:19
   |
62 |         outgoing: Option<(Multiaddr, Vec<(Multiaddr, TransportError<TTrans::Error>)>)>,
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::type_complexity)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_complexity

warning: very complex type used. Consider factoring parts into `type` definitions
  --> swarm/src/connection/pool.rs:93:18
   |
93 |       established: FnvHashMap<
   |  __________________^
94 | |         PeerId,
95 | |         FnvHashMap<
96 | |             ConnectionId,
97 | |             EstablishedConnection<<THandler::Handler as ConnectionHandler>::InEvent>,
98 | |         >,
99 | |     >,
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_complexity

warning: very complex type used. Consider factoring parts into `type` definitions
   --> swarm/src/connection/pool.rs:447:16
    |
447 |           dials: Vec<
    |  ________________^
448 | |             BoxFuture<
449 | |                 'static,
450 | |                 (
...   |
457 | |             >,
458 | |         >,
    | |_________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_complexity

warning: very complex type used. Consider factoring parts into `type` definitions
  --> swarm/src/connection.rs:77:21
   |
77 |       negotiating_in: FuturesUnordered<
   |  _____________________^
78 | |         SubstreamUpgrade<
79 | |             THandler::InboundOpenInfo,
80 | |             InboundUpgradeApply<SubstreamBox, SendWrapper<THandler::InboundProtocol>>,
81 | |         >,
82 | |     >,
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_complexity

warning: very complex type used. Consider factoring parts into `type` definitions
  --> swarm/src/connection.rs:84:22
   |
84 |       negotiating_out: FuturesUnordered<
   |  ______________________^
85 | |         SubstreamUpgrade<
86 | |             THandler::OutboundOpenInfo,
87 | |             OutboundUpgradeApply<SubstreamBox, SendWrapper<THandler::OutboundProtocol>>,
88 | |         >,
89 | |     >,
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_complexity

warning: very complex type used. Consider factoring parts into `type` definitions
   --> swarm/src/connection.rs:167:10
    |
167 |     ) -> Poll<Result<Event<THandler::OutEvent>, ConnectionError<THandler::Error>>> {
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_complexity

warning: very complex type used. Consider factoring parts into `type` definitions
   --> swarm/src/handler.rs:188:10
    |
188 |       ) -> Poll<
    |  __________^
189 | |         ConnectionHandlerEvent<
190 | |             Self::OutboundProtocol,
191 | |             Self::OutboundOpenInfo,
...   |
194 | |         >,
195 | |     >;
    | |_____^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_complexity

warning: `libp2p-swarm` (lib) generated 7 warnings
````

Reply:
Note that we have a custom set of clippy warnings, see https://github.com/libp2p/rust-libp2p/blob/master/.cargo/config.toml. Here the type complexity lint is ignored.

With that in mind, I suggest reverting the change in this pull request.

Reply:
Will do. TBH this was mostly for my own sake.

That said, I do think fine-tuning to ignore _just enough_ type complexity is better than turning off type complexity warnings in general.